Add passive beamloading cavity

lcarver opened the pull request:
Modified the existing functions to allow the user to specifiy a passive cavity. In this situation, Vgen is always zero, no RF cavity is tracked, only the vbeam option is tracked and applied to the particles. 

A new enumerator was provided called CavityMode and CavityMode.ACTIVE is the default.

It seems to be working well, but I don't have yet a good case to benchmark. So I have marked it as a WIP for the time being

lcarver commented:
Following the nice benchmarking results made by Teresia at the AT Workshop using this branch (slides are here and the good agreement pasted below: https://indico.esrf.fr/event/93/contributions/518/ ) 

I remove the WIP label and request review.

![image](https://github.com/atcollab/at/assets/16048082/cc97eab6-03fd-45ae-8dda-998bc343e605)

swhite2401 commented:
@lcarver, could you please update this branch with the master before review? The label WIP is still there too

lcarver commented:
> @lcarver, could you please update this branch with the master before review? The label WIP is still there too

done

swhite2401 commented on atintegrators/BeamLoadingCavityPass.c:
```diff
@@ -72,16 +74,20 @@ void BeamLoadingCavityPass(double *r_in,int num_particles,int nbunch,
 
-    /*Track RF cavity is always done. */
-    trackRFCavity(r_in,le,vgen/energy,rffreq,harmn,tlag,-psi,nturn,circumference/C0,num_particles);
+    if(cavitymode==2){
```
Here I would not change anything and integrate in trackRFCavity that if vgen=0 the loop over particle is skipped

lcarver replied:
done

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -13,7 +13,10 @@ class BLMode(IntEnum):
+    ACTIVE = 1
+    PASSIVE = 2
+    
 def add_beamloading(ring: Lattice, qfactor: Union[float, Sequence[float]],
```
Here you have to be careful depending on the mode you want vgen to be zero all the time: in build_from_cav and in init_bl_param you want this to be correctly handled

Unfortunately I can comment only the line that you changed...but lines 298 and 209 need to be adapted depending on the mode

Also do you want to allow setting Vgen for passive cavity? Should you return an error?

lcarver replied:
done

lcarver commented:
ready for re-review (and welcome to @TeresiaOlsson to the reviewing side of AT). 

I implemented the changes, now in the beamloading passmethod, if vgen is 0 and cav_length is 0 it skips the loop which means it can be in the main code irrespective of active or passive.
I added some checks to makes sure for a passive cavity vgen setpoint is always 0.

lfarv commented on pyat/at/collective/beam_loading.py:
```diff
@@ -139,12 +145,15 @@ def __init__(self, family_name: str, length: float, voltage: float,
             bl_elem (Element): beam loading element
         """
         kwargs.setdefault('PassMethod', self.default_pass[True])
         assert isinstance(mode, BLMode), \
             'Beam loading mode has to be an instance of BLMode'
+        assert isinstance(cavitymode, CavityMode), \
+            'Cavity mode has to be an instance of CavityMode'
```
As a general rule, you should avoid using `assert` for validity checks: `assert` statements are only executed when python is in debug mode, but  not in optimisation mode (python -O). Debug mode is currently the default, but it may change.
Instead, raise exceptions (TypeError in this case)

lcarver commented:
I have one other question to ask before I implement the changes (in case it's not needed).

At the moment I have to do 
mode=BLMode.WAKE, cavitymode=CavityMode.ACTIVE

should 'mode' be renamed 'blmode' in order to be clearer?

lcarver commented:
ready for review / merge

swhite2401 commented on atintegrators/BeamLoadingCavityPass.c:
```diff
@@ -72,16 +74,23 @@ void BeamLoadingCavityPass(double *r_in,int num_particles,int nbunch,
+    }else if(cavitymode==1){
+        vgen = vgenk[0];
+        psi = vgenk[1];
+    }
```
Here you may raise an error if the mode is unknown, or even in the initialization below where there are already sanity check

lcarver replied:
done

lcarver commented:
had to remerge master, i can't keep up!
also checked example files.

swhite2401 commented on atintegrators/attrackfunc.c:
```diff
@@ -15,11 +15,17 @@ void trackRFCavity(double *r_in, double le, double nv, double freq, double h, do
+        anyway if there is a cavity length, we have to loop through the particles
+    */
+
     if (le == 0) {
```
Here you can use
if(nv==0.0) return;
to exit the function instead. This would also cover the case with le!=0 which is not the case in your implementation

lcarver replied:
But if le is larger than 0 I don't want to exit the function, I want to apply the drift.

lfarv commented:
> had to remerge master, i can't keep up!

There is no need to merge master unless there is a conflict, which is not the case here. It just makes the commit history longer…

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -56,8 +63,9 @@ def apply(ring, cavpts, newelems):
-        assert isinstance(cav, RFCavity), \
-            'Beam loading can only be assigned to a cavity element'
+        if not isinstance(cav, RFCavity):
+            raise TypeError('Beam loading can only be assigned' +
```
raise AtError() and use Cavity element

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -90,8 +98,9 @@ def apply(ring, cavpts, newelems):
-        assert isinstance(bl, BeamLoadingElement), \
-            'Cannot remove beam loading: not a beam loading element'
+        if not isinstance(bl, BeamLoadingElement):
+            raise TypeError('Cannot remove beam loading:' +
```
raise AtError() and use BeamLoadingElement

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -135,16 +146,21 @@ def __init__(self, family_name: str, length: float, voltage: float,
-        assert isinstance(mode, BLMode), \
-            'Beam loading mode has to be an instance of BLMode'
+        if not isinstance(blmode, BLMode):
+            raise TypeError('Beam loading mode has to be an' +
```
raise AtError() and use blmode

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -135,16 +146,21 @@ def __init__(self, family_name: str, length: float, voltage: float,
+            raise TypeError('Beam loading mode has to be an' +
+                            'instance of BLMode')
+        if not isinstance(cavitymode, CavityMode):
+            raise TypeError('Cavity mode has to be an' +
```
raise AtError() and use cavitymode

lcarver replied:
Can you give an  example usage? AtError is fine but use cavitymode is the message?

swhite2401 replied:
Ah sorry, for this one and all the others it is better to put the argument name in the error message.
I would then put:
`raise TypeError('cavitymode has to be an instance of CavityMode')`

lcarver replied:
got it

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -193,16 +210,25 @@ def clear_history(self, ring=None):
+            psi = numpy.arcsin(b/numpy.sqrt(a**2+b**2))
+            vgen = self.Voltage*numpy.cos(psi) + \
+                vb*numpy.cos(psi)*numpy.sin(self._phis)
+
```
I am not very keen on these extra empty lines....

lcarver replied:
done

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -193,16 +210,25 @@ def clear_history(self, ring=None):
+        elif (self._cavitymode == 1) and (current == 0.0):
+            vgen = self.Voltage
+            psi = 0
+        else:
```
elif self._cavitymode==2

swhite2401 commented on pyat/at/collective/beam_loading.py:
```diff
@@ -193,16 +210,25 @@ def clear_history(self, ring=None):
+            vgen = self.Voltage*numpy.cos(psi) + \
+                vb*numpy.cos(psi)*numpy.sin(self._phis)
+
+        elif (self._cavitymode == 1) and (current == 0.0):
```
This should be an else at the end no?

lcarver replied:
I put an else and added an AtError for an unkown cavity and current configuration

swhite2401 replied:
Sorry I explained wrong, I was thinking of:

```
   def _init_bl_params(self, current):
        if (self._cavitymode == 1) and (current > 0.0):
            theta = -self._vcav[1]+numpy.pi/2
            vb = 2*current*self.Rshunt
            a = self.Voltage*numpy.cos(theta-self._phis)
            b = self.Voltage*numpy.sin(theta-self._phis)-vb*numpy.cos(theta)
            psi = numpy.arcsin(b/numpy.sqrt(a**2+b**2))
            vgen = self.Voltage*numpy.cos(psi) + \
                vb*numpy.cos(psi)*numpy.sin(self._phis)
        elif self._cavitymode == 2:
            vgen = 0
            psi = 0
        else:
            vgen = self.Voltage
            psi = 0   
```

lcarver commented:
changes implemented. Ready for re-review

swhite2401 commented on atintegrators/attrackfunc.c:
```diff
@@ -15,11 +15,17 @@ void trackRFCavity(double *r_in, double le, double nv, double freq, double h, do
+            }
         }
     }
     else {
```
Line 39
`if(nv!=0.0) r6[4] += -nv*sin(TWOPI*freq*((r6[5]-lag)/C0 - (h/freq-T0)*nturn) - philag);`

Would save you a sine calculation

lcarver replied:
done

lcarver commented:
@swhite2401 changes implemented

swhite2401 commented:
I all benchmarks are OK this is fine for me!

lcarver commented:
my local at directory is a bit of a mess right now as running checks for optimize_atimplib at the same time. I propose to wait until I am back next week, finalise some clean tests on this branch and then merge.

swhite2401 commented:
Yes ok better wait than merging something incorrect

lcarver commented:
I reran one of the MAXIV cases shown by Teresia (173 bunches, Nslice=50, 10k particles per bunch, 150k turns) and I was able to perfectly reproduce her results. So I am happy and will now merge.
